### ISSUE TITLE
Why does this snippet print "Hello World" endlessly? 

### ISSUE THREAD

**Issue by urev** [pkivolowitz/asm_book#4]
I have learned the `section_1/hello_world` and tested the code in a debian arm64 server.

This is the hello world snippet, `hello_v6.s` :

```
        .global _start
_start:
        str     x30, [sp, -16]!          // Preserve x30
        ldr     x0, =HW                 // Load address of string for puts
        bl      puts                        // puts(HW)
        ldr     x30, [sp], 16           // Restore x30
        mov     x0, xzr                 // return 0
        ret

        .data
HW:     .asciz  "Hello, World"
        .end
```

I compiled the code as follows:
```
as hello_v6.s -o hello_v6.o
ld hello_v6.o -lc
```

Then, run it:

```
./a.out
```

as --version:
```
GNU assembler (GNU Binutils for Debian) 2.35.2
Copyright (C) 2020 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or later.
This program has absolutely no warranty.
This assembler was configured for a target of `aarch64-linux-gnu'.
```

ld --version:
```
GNU ld (GNU Binutils for Debian) 2.35.2
Copyright (C) 2020 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

It prints "Hello World" endlessly. I thought only one "Hello World" would be printed. 

Would you like to tell me why?

**Comment by pkivolowitz** [comment]
_start is also a special name. It is what is called prior to main() completely behind the scenes.

May I ask, how did you choose this name?

[Here](https://embeddedartistry.com/blog/2019/04/08/a-general-overview-of-what-happens-before-main/) is a nice article (but Intel based).

Thanks

**Comment by pkivolowitz** [comment]
Was the preceding good for you? Shall I close this?
Thanks

**Comment by urev** [comment]
> _start is also a special name. It is what is called prior to main() completely behind the scenes.
> 
> May I ask, how did you choose this name?
> 
> [Here](https://embeddedartistry.com/blog/2019/04/08/a-general-overview-of-what-happens-before-main/) is a nice article (but Intel based).
> 
> Thanks

Because when I use `main`, the `ld` show me a warning:

```
ld: warning: cannot find entry symbol _start; defaulting to 00000000004002a0
```

Then, I found that changing the `main` into `_start` can suppress the warning.

**Comment by urev** [comment]
> Was the preceding good for you? Shall I close this? Thanks

Finally I found that this compiling method can solve the problem:

```
as hello_v6.s -o hello_v6
gcc hello_v6.o -o hello_v6
```

I checked the disassembly code between `hello_v4` and `hello_v6` and found the cause is that using the `_start` or `main` as the entry point. In the disassembly code of `hello_v4`, `_start` section does other things besides calling the `main` function.

Maybe I lost some options when using `ld` to do the link process.

Thank you for your reply and great tutorial!
